### PR TITLE
HPCC-26214 Check acceptable file path in ESP services

### PR DESCRIPTION
**Pull request by wangkx**
Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: esp/services/ws_topology/ws_topologyService.cpp [hunk 1]
```diff
@@ -236,6 +236,16 @@ bool CWsTopologyEx::onTpLogFile(IEspContext &context,IEspTpLogFileRequest  &req,
     return true;
 }
 
+void CWsTopologyEx::checkFilePath(const char *file, const char *fileType, const char *compType, const char *compName)
```
Comment (afishbeck) [inline esp/services/ws_topology/ws_topologyService.cpp:239]: minor, really just personal preference, and feel free to disagree, I would name the functions "validate" rather than "check" when their primary purpose is security related.

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp [hunk 2]
```diff
@@ -4011,6 +4011,19 @@ IFileIOStream* CWsWuFileHelper::createWUFileIOStream(IEspContext& context, const
     return createIOStreamWithFileName(zipFileNameWithPath.str(), IFOread);
 }
 
+void CWsWuFileHelper::checkFilePath(const char *file, bool UNCFileName, const char *fileType, const char *compType, const char *compName)
+{
+    StringBuffer actualPath;
+    if (UNCFileName)
```
Comment (afishbeck) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: You are splitting the path, validating one part, and then using the orinal path.  Is there any chance the other parts can be used to hack the system.  It may be the way they are parsed they won't have at least "../" but it's worth considering.

File: esp/services/ws_topology/ws_topologyService.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_topology/ws_topologyService.cpp:239]: I will change it.

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp (hunk 2 above)
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: The other parts are machine part (url or IP) and the tail part (the part after the last PathSepChar). Any idea how to check them?

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: The other parts are the machine part (url or IP) and the tail part (the part after the last PathSepChar). I haven't got an idea for the tail part. For the machine part, we may consider (in another PR?) to read the allowed machine list (with/without the allowed folder list for each machine) when the ESP is started (and when environment changes).

Comment (GordonSmith) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: If its available (c++17) you should probably run the full path through: 
 std::filesystem::canonical (https://en.cppreference.com/w/cpp/filesystem/canonical) as it will clean it up and give you the "real" path, including symlinks etc.

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: I just tried:
#include <iostream>
#include <filesystem>
namespace fs = std::filesystem;

I got: error: ‘filesystem’ is not a namespace-name

Comment (GordonSmith) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: Its ultimately a wrapper around realpath:  https://pubs.opengroup.org/onlinepubs/9699919799/functions/realpath.html

Which we do use already...

Comment (GordonSmith) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: Actually looks like jfile has wrapped it in `makeAbsolutePath`?

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: After using std::experimental::filesystem::path, I got:

/home/lexis/src3/HPCC-Platform/esp/services/ws_fs/ws_fsService.cpp:2800: undefined reference to `std::experimental::filesystem::v1::current_path[abi:cxx11]()'
/usr/bin/ld: /home/lexis/src3/HPCC-Platform/esp/services/ws_fs/ws_fsService.cpp:2800: undefined reference to `std::experimental::filesystem::v1::canonical(std::experimental::filesystem::v1::__cxx11::path const&, std::experimental::filesystem::v1::__cxx11::path const&)'
/usr/bin/ld: CMakeFiles/ws_fs.dir/ws_fsService.cpp.o: in function `std::experimental::filesystem::v1::__cxx11::path::path<char [16], std::experimental::filesystem::v1::__cxx11::path>(char const (&) [16])':
/usr/include/c++/9/experimental/bits/fs_path.h:212: undefined reference to `std::experimental::filesystem::v1::__cxx11::path::_M_split_cmpts()'
collect2: error: ld returned 1 exit status

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: I just tried the makeAbsolutePath:
```
    StringBuffer s1, s2, s3, s4;
    makeAbsolutePath("//10.0.1.15/a/b/../c/./d/e", s1);
    makeAbsolutePath("/a/b/../c/./d/e", s2);
    makeAbsolutePath("~/a/b/../c/./d/e", s3);
    makeAbsolutePath("\\x\\y\\..\\..\\a\\b\\c", s4);
    DBGLOG("####results s1=(%s), s2=(%s), s3=(%s), s4=(%s)", s1.str(), s2.str(), s3.str(), s4.str());
```
I got: "####results s1=(//10.0.1.15/a/b/../c/./d/e), s2=(/a/b/../c/./d/e), s3=(/home/lexis/a/b/../c/./d/e), `s4=(\x\y\..\..\a\b\c)"
I am not sure it helps...

Comment (GordonSmith) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: Hmmm the comments suggest it should be handling those...   If you debug it does it hit "realpath" at anypoint?

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: Yes for s2, s3, and s4.

Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4017]: After adding the link_libraries(stdc++fs) to CMakeLists, the std::experimental::filesystem::canonical works. But, it does not work for a remote file.

**Review by jakesmith — CHANGES_REQUESTED** [review]
@wangkx - please see comments

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 3]
```diff
@@ -2191,4 +2191,74 @@ extern TPWRAPPER_API unsigned getContainerWUClusterInfo(CConstWUClusterInfoArray
...
+        {
+            if (matchNetAddressRequest(netAddr, isIPAddressReq, tpMachines.item(ii)))
+                return true;
+        }
+    }
+#else
+    Owned<IEnvironmentFactory> envFactory = getEnvironmentFactory(true);
+    Owned<IConstEnvironment> constEnv = envFactory->openEnvironment();
+    Owned<IConstDropZoneInfoIterator> dropZoneItr = constEnv->getDropZoneIteratorByAddress(netAddr);
+    ForEach(*dropZoneItr)
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/smc/SMCLib/TpWrapper.cpp:2240]: Not for7.12.x. That change was introduced to 8.0.0 by https://github.com/hpcc-systems/HPCC-Platform/commit/18ea181ac341f83c7c75258e1dd0a24b3ef4451f.

File: esp/services/ws_fileio/ws_fileioservice.cpp [hunk 4]
```diff
@@ -37,6 +37,9 @@ bool CWsFileIOEx::CheckServerAccess(const char* targetDZNameOrAddress, const cha
     if (!targetDZNameOrAddress || (targetDZNameOrAddress[0] == 0) || !relPath || (relPath[0] == 0))
         return false;
 
+    if (hasSpecialPathCharacters(relPath)) //Detect a path like: a/../../../f
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fileio/ws_fileioservice.cpp:40]: I guess I have to check the char after a dot.

[the remark this answers is not in the document]
Comment (jakesmith) [reply esp/services/ws_fileio/ws_fileioservice.cpp:40]: it's more that it/they appear on it's own after a path separator

[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fileio/ws_fileioservice.cpp:40]: Right. Is it enough just to check the char before the dot?

[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fileio/ws_fileioservice.cpp:40]: If the dot is the first char, it also should be invalid.

**Comment by wangkx** [comment]
@jakesmith I made the changes based on your comments. Please review.

File: system/jlib/jfile.cpp [hunk 5]
```diff
@@ -5252,12 +5252,18 @@ bool hasSpecialPathCharacters(const char *path)
...
+    char sepChar = getPathSepChar(path);
+    char previousChar = '\0';
     const char *cur = path;
     for (;;)
     {
-        switch (*cur++)
+        switch (*cur)
         {
         case '.':
+            if ((previousChar=='\0') || (previousChar==sepChar)) //the dot is the frist char or after a path separator.
```
Comment (jakesmith) [inline system/jlib/jfile.cpp:5264]: I think ok, in the sense that the context hasSpecialPathCharacters() is used it's unlikely to trip up,
but this could still produce false negatives, e.g. /.mydir/ is a valid path (& and % are valid chars really too).

If the main objective is to spot relative paths, perhaps this method should be renamed (e.g. containsRelPaths), and it explitly walks and looks for only '~', '.' and '..'

Comment (wangkx) [reply system/jlib/jfile.cpp:5264]: Ok, I will: 1. rename; 2. only looks for '~', '.' and '..'; 3. check the char after a dot.

Comment (wangkx) [reply system/jlib/jfile.cpp:5264]: Is the /..mydir/ a valid path?

Comment (jakesmith) [reply system/jlib/jfile.cpp:5264]: yes "." and ".." only mean anything if they're a path directive, leading/trailing or between path separators.

Comment (jakesmith) [reply system/jlib/jfile.cpp:5264]: and technically '~' is only meaningful as leading, e.g. really 'a~b' is a legal posix path/filename

**Comment by wangkx** [comment]
@jakesmith please review

**Review by jakesmith — APPROVED** [review]
Looks ok to me.
@afishbeck - could you give it one more look please

**Review by afishbeck — APPROVED** [review]
@wangkx yes, looks fine to me.

**Comment by wangkx** [comment]
@richardkchapman this PR has been approved.

**Review by afishbeck — APPROVED** [review]
Additional change looks fine.

**Comment by wangkx** [comment]
I just did squash.

File: esp/services/ws_fs/ws_fsService.cpp [hunk 6]
```diff
@@ -2815,6 +2815,9 @@ bool CFileSprayEx::onFileList(IEspContext &context, IEspFileListRequest &req, IE
                 throw MakeStringException(ECLWATCH_ACCESS_TO_FILE_DENIED, "Only cfg or log file allowed.");
         }
 
+        if (!validateDropZonePath(nullptr, netaddr, sPath) && !validateConfigurationDirectory(nullptr, "log", nullptr, nullptr, sPath)) //The path should be the absolute path for the dropzone or log file.
```
Comment (ghalliday) [inline esp/services/ws_fs/ws_fsService.cpp:2818]: I know this has generally been the case, but requiring an absolute name for the dropzone or log path seems wrong - since it is exposing internal information that shouldn't be required.  Should probably be revisited in a separate PR

File: esp/services/ws_topology/ws_topologyService.cpp [hunk 7]
```diff
@@ -236,6 +236,16 @@ bool CWsTopologyEx::onTpLogFile(IEspContext &context,IEspTpLogFileRequest  &req,
     return true;
 }
 
+void CWsTopologyEx::validateFilePath(const char *file, const char *fileType, const char *compType, const char *compName)
+{
+    StringBuffer actualPath;
+    splitUNCFilename(file, nullptr, &actualPath, nullptr, nullptr);
+    if (containsRelPaths(actualPath)) //Detect a path like: /home/lexis/runtime/var/log/HPCCSystems/myesp/../../../
+        throw makeStringExceptionV(ECLWATCH_INVALID_INPUT, "Invalid file path %s", actualPath.str());
+    if (!validateConfigurationDirectory(nullptr, fileType, compType, compName, actualPath))
```
Comment (ghalliday) [inline esp/services/ws_topology/ws_topologyService.cpp:245]: This is only going to work for log files if there isn't an override for a particular instance in the environment file, and then also only works because [INSTANCE] is the trailing component.  It probably works for most existing environment files, but not for all.

Comment (ghalliday) [inline esp/services/ws_topology/ws_topologyService.cpp:245]: Potential performance issues because this will reread environment.xml each time it is called?  (getOSSdirTree() is not cached.)

File: esp/scm/ws_topology.ecm [hunk 8]
```diff
@@ -575,6 +576,7 @@ ESPrequest [nil_remove] SystemLogRequest
     string Name;
     string Type;
     int   Zip;
+    [min_ver("1.32")] string CompType;
```
Comment (ghalliday) [inline esp/scm/ws_topology.ecm:579]: This should have some kind of documentation.  Does eclwatch need to be updated to provide this argument, or will it mean that the legacy SystemLog option will no longer work.

File: esp/scm/ws_topology.ecm [hunk 9]
```diff
@@ -543,6 +543,7 @@ ESPrequest [nil_remove] TpLogFileRequest
     [min_ver("1.05")] int PageNumber(0);
     [min_ver("1.05")] bool LoadData(false);
     [min_ver("1.29")] bool IncludeLogFieldNames(true);
+    [min_ver("1.32")] string CompType;
```
Comment (ghalliday) [inline esp/scm/ws_topology.ecm:546]: This will require changes to eclwatch (e.g. WsTopology.ts, tplog.xslt) for it to work.

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp [hunk 10]
```diff
@@ -4028,11 +4041,18 @@ void CWsWuFileHelper::readWUFile(const char* wuid, const char* workingFolder, Ws
     case CWUFileType_CPP:
     case CWUFileType_LOG:
     {
-        const char *tail=pathTail(item.getName());
-        fileName.set(tail ? tail : item.getName());
+        const char *file=item.getName();
+#ifndef _CONTAINERIZED
+        validateFilePath(file, false, "run", nullptr, nullptr);
+#else
+        validateFilePath(file, false, "query", nullptr, nullptr);
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp:4048]: This is potentially not going to work for containerized systems - you can have c++ files and dlls stored in multiple locations.  An unusual case though.
The idea that you can get a single location and then compare it as a prefix is likely to have other instances (like log overrides in bare metal) where it doesn't really work.

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp [hunk 11]
```diff
@@ -4052,11 +4072,16 @@ void CWsWuFileHelper::readWUFile(const char* wuid, const char* workingFolder, Ws
         writeToFileIOStream(workingFolder, fileName.str(), mb);
         break;
     case CWUFileType_ThorLog:
+    {
+        const char *file=item.getName();
+        validateFilePath(file, true, "log", nullptr, nullptr);
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp:4077]: doesn't this need a component type?

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp [hunk 12]
```diff
@@ -4068,11 +4093,16 @@ void CWsWuFileHelper::readWUFile(const char* wuid, const char* workingFolder, Ws
         break;
     }
     case CWUFileType_EclAgentLog:
+    {
+        const char *file=item.getName();
+        validateFilePath(file, true, "log", nullptr, nullptr);
```
Comment (ghalliday) [inline esp/services/ws_workunits/ws_workunitsHelpers.cpp:4098]: also requires a component type?

**Review by ghalliday — COMMENTED** [review]
@wangkx a few comments.  I've been very picky since this is being merged into 7.12.x.
Especially looking for places where this will break existing functionality.  I think on fairly standard configurations it should work, but weird environment.xml file mappings are not going to work.
Why is CompType added to the topology calls?  If it is required there are quite a few other calls that are missing it.
The general assumption that you can get a single directory, and then check it is a prefix is likely to not work in various contexts.

File: esp/scm/ws_topology.ecm (hunk 8 above)
Comment (wangkx) [reply esp/scm/ws_topology.ecm:579]: The CompType is used to set the component variable inside the getConfigurationDirectory(). So far, I haven't found which ConfigurationDirectory needs that variable. I add the CompType here just in case we need it some time later. Do you know when the component variable has to be specified?

File: esp/services/ws_topology/ws_topologyService.cpp (hunk 7 above)
Comment (wangkx) [reply esp/services/ws_topology/ws_topologyService.cpp:245]: Do we know where those special log file could be located? Should we add some settings to specify the paths to be allowed for retrieving the log files (and force users to modify their environment.xml)?

Comment (wangkx) [reply esp/services/ws_topology/ws_topologyService.cpp:245]: Do you know why getOSSdirTree() is not cached?

File: esp/services/ws_fs/ws_fsBinding.cpp [hunk 13]
```diff
@@ -470,6 +473,17 @@ int CFileSpraySoapBindingEx::downloadFile(IEspContext &context, CHttpRequest* re
     return 0;
 }
 
+int CFileSpraySoapBindingEx::onStartUpload(IEspContext& ctx, CHttpRequest* request, CHttpResponse* response, const char* serv, const char* method)
+{
+    StringBuffer netAddress, path;
+    request->getParameter("NetAddress", netAddress);
+    request->getParameter("Path", path);
+    if (!validateDropZonePath(nullptr, netAddress, path)) //The path should be the absolute path for the dropzone.
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fs/ws_fsBinding.cpp:481]: That code in the EspHttpBinding::onStartUpload is for adding package, not for dropzone file.

File: system/jlib/jutil.cpp [hunk 14]
```diff
@@ -2715,6 +2715,22 @@ bool replaceConfigurationDirectoryEntry(const char *path,const char *frommask,co
...
+{
+    if (isEmptyString(dirToValidate))
+        return false;
+
+    StringBuffer configDir;
+    if (!getConfigurationDirectory(useTree, category, component, instance, configDir))
+        return false;
+    
+    addPathSepChar(configDir);
+    unsigned configDirLen = configDir.length();
```
[the remark this answers is not in the document]
Comment (wangkx) [reply system/jlib/jutil.cpp:2728]: Good to know. I will change.

File: system/jlib/jfile.cpp [hunk 15]
```diff
@@ -5248,6 +5248,47 @@ bool isRemotePath(const char *path)
...
+            break;
+        }
+        case '\0':
+            return false;
+        }
+
+        previousChar = *cur;
+        cur++;
+    }
+    return false;
```
[the remark this answers is not in the document]
Comment (wangkx) [reply system/jlib/jfile.cpp:5289]: Will remove.

File: esp/smc/SMCLib/TpWrapper.cpp [hunk 16]
```diff
@@ -2191,4 +2191,74 @@ extern TPWRAPPER_API unsigned getContainerWUClusterInfo(CConstWUClusterInfoArray
...
+{
+    if (ipReq)
+        return streq(netAddressReg, tpMachine.getNetaddress());
+    return streq(netAddressReg, tpMachine.getConfigNetaddress());
+}
+
+extern TPWRAPPER_API bool validateDropZonePath(const char* dropZoneName, const char* netAddr, const char* pathToCheck)
+{
+    if (isEmptyString(netAddr) || isEmptyString(pathToCheck))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "NetworkAddress or Path not defined.");
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/smc/SMCLib/TpWrapper.cpp:2206]: Will change

File: esp/services/ws_fs/ws_fsBinding.hpp [hunk 17]
```diff
@@ -75,6 +75,7 @@ class CFileSpraySoapBindingEx : public CFileSpraySoapBinding
     }
 
     int onGetInstantQuery(IEspContext &context, CHttpRequest* request, CHttpResponse* response, const char *service, const char *method);
+    int onStartUpload(IEspContext& ctx, CHttpRequest* request, CHttpResponse* response, const char* serv, const char* method);
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fs/ws_fsBinding.hpp:78]: I will add them.

File: esp/scm/ws_topology.ecm (hunk 8 above)
Comment (ghalliday) [reply esp/scm/ws_topology.ecm:579]: My conclusion was for the default environment file none of default definitions need it to be set (because it is the trailing part of the file path).  Someone could define a strange environment where all cases will need it to be defined.  (It would also be required if a component-specific override was defined.)
ComponentType would be a better name than CompType, slightly clearer what it corresponds to.

File: esp/services/ws_fs/ws_fsBinding.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (ghalliday) [reply esp/services/ws_fs/ws_fsBinding.cpp:481]: I asked because you are calling through to that base definition, which suggests it is the function that would have been called previously.

File: esp/scm/ws_topology.ecm (hunk 9 above)
Comment (wangkx) [reply esp/scm/ws_topology.ecm:546]: I haven't found a place where the CompType has to be set so far.

File: esp/services/ws_topology/ws_topologyService.cpp (hunk 7 above)
Comment (ghalliday) [reply esp/services/ws_topology/ws_topologyService.cpp:245]: Because no one has implemented it.  I noticed, and nearly add caching when I was modifying code in that area.  It needs to be a time limited cache - and should be a separate PR.

File: esp/scm/ws_topology.ecm (hunk 8 above)
Comment (ghalliday) [reply esp/scm/ws_topology.ecm:579]: If it is not specified then the user will be able to get to any logs.

File: esp/services/ws_topology/ws_topologyService.cpp (hunk 7 above)
Comment (ghalliday) [reply esp/services/ws_topology/ws_topologyService.cpp:245]: They could theoretically be located anywhere.  I don't know if any real-life environments actually hit this.

**Review by ghalliday — COMMENTED** [review]
@wangkx a few responses.
I'm not sure what should be done about component type.  For a default environment I don't think it is needed.  To cope with all environments you would need a different approach since you can also have exceptions for particular instances (generate a list of all potential paths and check if it matches any of them).
I'm not sure if adding CompType is worthwhile - since it raises the question of why it isn't set in other equivalent calls.

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp (hunk 11 above)
Comment (ghalliday) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4077]: ignore - see later comment

File: system/jlib/jfile.cpp [hunk 18]
```diff
@@ -5248,6 +5248,47 @@ bool isRemotePath(const char *path)
...
 
+bool containsRelPaths(const char *path)
+{
+    if (isEmptyString(path))
+        return false;
+    if (*path == '~')
+        return true;
+
+    const char *cur = path;
+    char sepChar = getPathSepChar(path);
```
[the remark this answers is not in the document]
Comment (wangkx) [reply system/jlib/jfile.cpp:5259]: How about I add a getPathSepCharEx() to implement this logic?

File: esp/services/ws_fs/ws_fsBinding.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_fs/ws_fsBinding.cpp:481]: Sorry, I think I did not explain it clearly. When adding a package, the CWsPackageProcessSoapBindingEx (not CFileSpraySoapBindingEx) is used. The onStartUpload is not overrided in CWsPackageProcessSoapBindingEx.  So, the EspHttpBinding::onStartUpload will be called directly for adding package. Both NetAddress and Path are empty when adding package.

File: esp/services/ws_topology/ws_topologyService.cpp (hunk 7 above)
Comment (wangkx) [reply esp/services/ws_topology/ws_topologyService.cpp:245]: As I know, the function here is used for accessing the configuration files and log files of HPCC components (most likely through ECLWatch). Any suggestion about how to do with those special log files?

File: esp/services/ws_workunits/ws_workunitsHelpers.cpp (hunk 10 above)
Comment (wangkx) [reply esp/services/ws_workunits/ws_workunitsHelpers.cpp:4048]: Any suggestion about how to do with the c++ files and dlls stored in multiple locations?

**Comment by wangkx** [comment]
Since the CompType raises the question and no use case has been found yet, I should remove it for this PR.

**Comment by ghalliday** [comment]
@richardkchapman what do you think?
I am inclined to agree.  We should either pass null in all places, or allow the comp to be supplied in all places - currently it is a bit of a mixture.  With the standard environment it makes no difference since the component is the trailing entry on the path.  With a completely custom environment this solution may not work anyway (it would also need the instance).  It also potentially has issues with different instances storing information in different locations (unusual k8s configurations).

I would remove comp from the services, and accept that this change does a reasonable job on standard environments, but could fail on custom environments.  Possibly revisit the implementation if multiple dll storage planes (or something else that isn't supported) turns out to be something that is used.

**Comment by wangkx** [comment]
@ghalliday please review. If I miss anything, please let me know.

File: esp/services/ws_topology/ws_topologyService.cpp (hunk 7 above)
Comment (ghalliday) [reply esp/services/ws_topology/ws_topologyService.cpp:245]: Created HPCC-26316

File: system/jlib/jfile.cpp [hunk 19]
```diff
@@ -5248,6 +5248,46 @@ bool isRemotePath(const char *path)
...
+    for (;;)
+    {
+        switch (*cur)
+        {
+        case '.':
+        {
+            if (previousChar == sepChar)
+            {
+                cur++;
+                if ((*cur == sepChar) || (*cur == '\0')) // '.'
```
Comment (ghalliday) [inline system/jlib/jfile.cpp:5270]: @jakesmith should this be disallowed?  @GordonSmith could this cause any problems?
It means \var\lib\HPCCSystems\myhthor\.\hthor.log
will be disallowed.  Could there be any existing code where .\hthor.log is passed in as a relative path?  (Not just for logs but any of these affected functions.)

File: system/jlib/jfile.hpp [hunk 20]
```diff
@@ -491,6 +491,19 @@ inline char getPathSepChar(const char *dir)
     return s?(*s):((*dir&&(dir[1]==':'))?'\\':PATHSEPCHAR);
 }
 
+inline char getPathSepCharEx(const char *s)
+{
+    bool foundBackslash=false;
+    if (s) while (*s) {
```
Comment (ghalliday) [inline system/jlib/jfile.hpp:497]: I see that the code is cloned and modified, but...
non standard nesting of if/while and bracket placement.  Probably better with the definition in the c++ file (reduces compile times and complication of the header.

**Review by ghalliday — COMMENTED** [review]
@wangkx a couple of minor comments.  I think it would be ok to merge as-is, but I would like to double check that /./ will not cause problems, and just about worth cleaning up the jfile function.
